### ISSUE TITLE
[8.x] After upgrade to PHP 8.0 transactions stop to work on seeders

### ISSUE THREAD

**Issue by eusonlito** [laravel/framework#35161]
- Laravel Version: 8.13.0
- PHP Version: 8.0.0RC3
- Database Driver & Version: MySQL 8.0.22

### Description:

Executing seeders with transactions on PHP 8.0 are failing with exception:

```
$> php8.0 artisan migrate:refresh && php8.0 artisan db:seed --class=Database\\Seeds\\Configuration

Migration table not found.
Migration table created successfully.
Migrating: 2020_02_05_000000_base
Migrated:  2020_02_05_000000_base (1,408.24ms)

   PDOException 

  There is no active transaction

  at vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php:44
     40▕             }
     41▕ 
     42▕             try {
     43▕                 if ($this->transactions == 1) {
  ➜  44▕                     $this->getPdo()->commit();
     45▕                 }
     46▕ 
     47▕                 $this->transactions = \max(0, $this->transactions - 1);
     48▕             } catch (Throwable $e) {

      +3 vendor frames 
  4   database/Seeds/Configuration.php:17
      Illuminate\Support\Facades\Facade::__callStatic()

      +22 vendor frames 
  27  artisan:37
      Illuminate\Foundation\Console\Kernel::handle()
```

Same command with PHP 7.4 is working as expected:

```
$> php7.4 artisan migrate:refresh && php7.4 artisan db:seed --class=Database\\Seeds\\Configuration

Migration table not found.
Migration table created successfully.
Migrating: 2020_02_05_000000_base
Migrated:  2020_02_05_000000_base (1,432.37ms)
Database seeding completed successfully.
```

### Steps To Reproduce:

1. Create any seeder like:

```php
<?php declare(strict_types=1);

namespace Database\Seeds;

use Illuminate\Database\Seeder;
use Illuminate\Support\Facades\DB;

class Configuration extends Seeder
{
    /**
     * @return void
     */
    public function run()
    {
        DB::transaction(function () {
            DB::table('configuration')->truncate();
        });
    }
}
```

2. Execute seeder

**Comment by eusonlito** [comment]
Some fast debug:

File: `src/Illuminate/Database/Concerns/ManagesTransactions.php`:
Method: `createTransaction`
Line: `126`

```php
dump($this->getPdo()->beginTransaction());
```

```
true
```

File: `src/Illuminate/Database/Concerns/ManagesTransactions.php`:
Method: `transaction`
Line: `44`

```php
dump($this->getPdo()->inTransaction());
```

```
false
```

**Comment by GrahamCampbell** [comment]
Did you upgrade any of your other dependencies at the same time? It would be good know if the exact same dependencies work on PHP 7.4, but not 8.0.

**Comment by eusonlito** [comment]
No, anything was upgraded, only executed the commands with one and other version. I will check now with a previous composer install with PHP 8.0.

**Comment by eusonlito** [comment]
Same error after `composer update` or `composer install` with PHP 8.0.

**Comment by GrahamCampbell** [comment]
Are you able to post your `composer.lock` file?

**Comment by eusonlito** [comment]
Tested on a new Laravel installation with same error:

```
$> composer create-project --prefer-dist laravel/laravel
$> cd laravel
$> vi .env # Database configuration
$> mysql -e 'CREATE DATABASE `test` character set UTF8mb4 collate utf8mb4_bin;'
$> cp database/migrations/2014_10_12_000000_create_users_table.php database/migrations/2014_10_12_000000_create_configuration.php
$> vi database/migrations/2014_10_12_000000_create_configuration.php # Edited class name and table name to "configuration"
$> vi database/seeders/Configuration.php # Created seeder with updated namespace Database\Seeds > Database\Seeders
$> composer dump-autoload
$> php8.0 artisan migrate:refresh && php8.0 artisan db:seed --class=Database\\Seeders\\Configuration

Rolling back: 2019_08_19_000000_create_failed_jobs_table
Rolled back:  2019_08_19_000000_create_failed_jobs_table (31.07ms)
Rolling back: 2014_10_12_100000_create_password_resets_table
Rolled back:  2014_10_12_100000_create_password_resets_table (17.60ms)
Rolling back: 2014_10_12_000000_create_users_table
Rolled back:  2014_10_12_000000_create_users_table (15.77ms)
Migrating: 2014_10_12_000000_create_configuration
Migrated:  2014_10_12_000000_create_configuration (75.21ms)
Migrating: 2014_10_12_000000_create_users_table
Migrated:  2014_10_12_000000_create_users_table (73.39ms)
Migrating: 2014_10_12_100000_create_password_resets_table
Migrated:  2014_10_12_100000_create_password_resets_table (70.18ms)
Migrating: 2019_08_19_000000_create_failed_jobs_table
Migrated:  2019_08_19_000000_create_failed_jobs_table (70.02ms)

   PDOException 

  There is no active transaction

  at vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php:44
     40▕             }
     41▕ 
     42▕             try {
     43▕                 if ($this->transactions == 1) {
  ➜  44▕                     $this->getPdo()->commit();
     45▕                 }
     46▕ 
     47▕                 $this->transactions = max(0, $this->transactions - 1);
     48▕             } catch (Throwable $e) {

      +3 vendor frames 
  4   database/seeders/Configuration.php:17
      Illuminate\Support\Facades\Facade::__callStatic()

      +22 vendor frames 
  27  artisan:37
      Illuminate\Foundation\Console\Kernel::handle()
```

`composer.lock` is too large to paste here: https://pastebin.pl/view/4dadccc3

Thanks!

**Comment by eusonlito** [comment]
Full debug on new installation:

```
[2020-11-10 07:07:01] local.ERROR: There is no active transaction {"exception":"[object] (PDOException(code: 0): There is no active transaction at /home/www/laravel/vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php:44)
[stacktrace]
#0 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php(44): PDO->commit()
#1 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Database/DatabaseManager.php(367): Illuminate\\Database\\Connection->transaction()
#2 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(261): Illuminate\\Database\\DatabaseManager->__call()
#3 /home/www/laravel/database/seeders/Configuration.php(17): Illuminate\\Support\\Facades\\Facade::__callStatic()
#4 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Database\\Seeders\\Configuration->run()
#5 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/Util.php(40): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#6 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#7 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod()
#8 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/Container.php(596): Illuminate\\Container\\BoundMethod::call()
#9 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Database/Seeder.php(149): Illuminate\\Container\\Container->call()
#10 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Database/Console/Seeds/SeedCommand.php(65): Illuminate\\Database\\Seeder->__invoke()
#11 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/GuardsAttributes.php(157): Illuminate\\Database\\Console\\Seeds\\SeedCommand->Illuminate\\Database\\Console\\Seeds\\{closure}()
#12 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Database/Console/Seeds/SeedCommand.php(66): Illuminate\\Database\\Eloquent\\Model::unguarded()
#13 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Illuminate\\Database\\Console\\Seeds\\SeedCommand->handle()
#14 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/Util.php(40): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#15 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#16 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod()
#17 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Container/Container.php(596): Illuminate\\Container\\BoundMethod::call()
#18 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Console/Command.php(136): Illuminate\\Container\\Container->call()
#19 /home/www/laravel/vendor/symfony/console/Command/Command.php(258): Illuminate\\Console\\Command->execute()
#20 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Console/Command.php(121): Symfony\\Component\\Console\\Command\\Command->run()
#21 /home/www/laravel/vendor/symfony/console/Application.php(920): Illuminate\\Console\\Command->run()
#22 /home/www/laravel/vendor/symfony/console/Application.php(266): Symfony\\Component\\Console\\Application->doRunCommand()
#23 /home/www/laravel/vendor/symfony/console/Application.php(142): Symfony\\Component\\Console\\Application->doRun()
#24 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Console/Application.php(93): Symfony\\Component\\Console\\Application->run()
#25 /home/www/laravel/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(129): Illuminate\\Console\\Application->run()
#26 /home/www/laravel/artisan(37): Illuminate\\Foundation\\Console\\Kernel->handle()
#27 {main}
```

Also I have removed all try/catch on ManagesTransactions.php and Connection.php to avoid exceptions catch noise and same error.

**Comment by eusonlito** [comment]
It's strange, error should be the correct behaviour in MySQL: https://dev.mysql.com/doc/refman/8.0/en/truncate-table.html

> Truncate operations cause an implicit commit, and so cannot be rolled back. See Section 13.3.3, “Statements That Cause an Implicit Commit”.

But I don't know why only fails with a different PHP version and not MySQL. Maybe some PHP PDO bug on 7.x?

I have added more debug into `transaction` method:

```php
    public function transaction(Closure $callback, $attempts = 1)
    {
        for ($currentAttempt = 1; $currentAttempt <= $attempts; $currentAttempt++) {
            $this->beginTransaction();

            // We'll simply execute the given callback within a try / catch block and if we
            // catch any exception we can rollback this transaction so that none of this
            // gets actually persisted to a database or stored in a permanent fashion.
            try {
                dump(__METHOD__.'::'.__LINE__.'::inTransaction', $this->getPdo()->inTransaction());
                $callbackResult = $callback($this);
                dump(__METHOD__.'::'.__LINE__.'::inTransaction', $this->getPdo()->inTransaction());
            }
```

With PHP 7.4:

```
$> php7.4 artisan db:seed --class=Database\\Seeders\\Configuration

"Illuminate\Database\Concerns\ManagesTransactions::transaction::28::inTransaction"
true
"Illuminate\Database\Concerns\ManagesTransactions::transaction::30::inTransaction"
true
Database seeding completed successfully.
```

With PHP 8.0:

```
$> php8.0 artisan db:seed --class=Database\\Seeders\\Configuration

"Illuminate\Database\Concerns\ManagesTransactions::transaction::28::inTransaction"
true
"Illuminate\Database\Concerns\ManagesTransactions::transaction::30::inTransaction"
false

   PDOException 

  There is no active transaction

  at vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php:46
     42▕             }
     43▕ 
     44▕             try {
     45▕                 if ($this->transactions == 1) {
  ➜  46▕                     $this->getPdo()->commit();
     47▕                 }
     48▕ 
     49▕                 $this->transactions = max(0, $this->transactions - 1);
     50▕             } catch (Throwable $e) {

      +3 vendor frames 
  4   database/seeders/Configuration.php:17
      Illuminate\Support\Facades\Facade::__callStatic()

      +22 vendor frames 
  27  artisan:37
      Illuminate\Foundation\Console\Kernel::handle()
```

As not a Laravel issue, I close this :)

Thanks @GrahamCampbell for your interest.

**Comment by litan1106** [comment]
I ran into this no active transaction issue with PHP 8.0 and mysql too.

**Comment by crynobone** [comment]
<img width="742" alt="Screenshot 2020-12-04 at 2 51 51 PM" src="https://user-images.githubusercontent.com/172966/101131442-4518eb80-3640-11eb-8b91-c456fcf0c471.png">

Laravel docs has been updated to reflect this issue.

**Comment by driesvints** [comment]
@litan1106 also: this isn't a breaking change per se. The issue was always there. PHP 8 just makes it more explicit now.

**Comment by telkins** [comment]
I don't mean to resurrect a dead issue, but I am wondering if there may be something wrong...still/now.

I'm using the latest Laravel (with Jetstream/Fortify).

I'm using statements that cause implicit commits....but I'm doing them in jobs (or listeners) using `afterCommit()`...and I'm still getting the "There is no active transaction" error message.

I've tried to isolate it and have done a fairly decent job, and I'm executing these jobs when teams are created or deleted.  It's very frustrating and I've got something that seems to work, but it's quite hackish, imho, and ought not be necessary.

Anyway, I know I might be doing something wrong, but I'd still like some input/feedback from people who are doing more or less the same kind of thing I'm doing and either struggling or who seem to have found success.

Again, I'm basically creating/destroying a database for each (jetstream) team that's created.  The main events are user creation, team creation, team deletion, and user deletion.

I'd love some feedback/input.  🤓
